content/blog/taksim-modes-makam.md

```diff
@@ -0,0 +1,664 @@
+Is there some way that we can extend Makam to support this recipe?
+
+## modal makam
+
+```makam-hidden
+(* stdlib extensions required for this. will open a PR for Makam soon *)
+
+%extend refl.
+
+decompose_term : [Full Head] Full -> Head -> args Head Full -> prop.
+decompose_term Term Head Args :-
+  refl.headargs Term Head ArgsDyn,
+  dyn.to_args ArgsDyn Args.
+
+recompose_term : [Full Head] Head -> args Head Full -> Full -> prop.
+recompose_term Head Args Term :-
+  args.applyfull Head Args Term.
+
+%end.
+```
+
+```makam-hidden
+%extend modes_v1.
+```
+
+Here's a first approach: we can do some meta-programming in Makam, transforming predicates that utilize a specific mode according to the recipe. We will have to perform that transformation for all rules of the relevant predicates.
+
+Let's see how this would work. First of all, we can define `mode` as a predicate, that will be used to specify the mode for predicates. We'll make light use of GADTs to make sure that when specifying a mode, the number of arguments we specify matches the number of arguments of the predicate.
+
+```makam
+argument_mode : type.
+input, output : argument_mode.
+
+mode_spec : type -> type.
+nil : mode_spec prop.
+cons : argument_mode -> mode_spec B -> mode_spec (A -> B).
+
+mode : [A] A -> mode_spec A -> prop.
+```
+
+Now we should be able to specify modes for our predicates:
+
+```makam
+type_check : expr -> typ -> prop.
+mode type_check [input, input].
+
+type_synth : expr -> typ -> prop.
+mode type_synth [input, output].
+```
+
+Now let's work on the actual clause transformation. A clause like the following:
+
+```makam-noeval
+type_check (lambda E) (arrow A1 A2) :-
+  (x:expr -> type_of_var x A1 -> type_check (E x) A2).
+```
+
+should be turned to:
+
+```makam-noeval
+type_check Arg1 Arg2 :-
+  pattern_match (lambda E) Arg1,
+  pattern_match (arrow A1 A2) Arg2,
+  (x:expr -> type_of_var x A1 -> type_check (E x) A2).
+```
+
+Desugaring, the above clauses correspond to the following Makam terms:
+
+```makam-noeval
+clause (type_check (lambda E) (arrow A1 A2))
+  (newvar (fun (x:expr) =>
+    (assume (clause (type_of_var x A1) success)
+      (type_check (E x) A2))))
+```
+
+and after transformation:
+
+```makam-noeval
+clause (type_check Arg1 Arg2)
+  (and_many
+    [ pattern_match (lambda E) Arg1,
+      pattern_match (arrow A1 A2) Arg2,
+      newvar (fun (x:expr) =>
+        (assume (clause (type_of_var x A1) success)
+          (type_check (E x) A2))) ])
+```
+
+Let's write a predicate to do that transformation:
+
+```makam
+generate_matches :
+  [A] mode_spec A ->
+      args A prop -> args A prop -> list prop -> prop.
+
+mode_transform : clause -> clause -> prop.
+mode_transform
+    (clause Goal Premise)
+    (clause Goal' (and_many MatchesAndPremise)) :-
+  refl.decompose_term Goal Predicate Patterns,
+  refl.recompose_term Predicate Args Goal',
+  mode Predicate ModeSpec,
```

Comment:
Oh, is this where things are tied together? The fact that `mode` never gets any clauses defined for it is a bit of a leap for me 😅